// typescript-language-server/lib/ts-protocol.js
/*
 * Copyright (C) 2018 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * **IMPORTANT** this module should not depend on `vscode-languageserver` only protocol and types
 */
import lsp from 'vscode-languageserver-protocol';
export var TypeScriptRenameRequest;
(function (TypeScriptRenameRequest) {
    TypeScriptRenameRequest.type = new lsp.RequestType('_typescript.rename');
})(TypeScriptRenameRequest = TypeScriptRenameRequest || (TypeScriptRenameRequest = {}));
export class DisplayPartKind {
}
DisplayPartKind.functionName = 'functionName';
DisplayPartKind.methodName = 'methodName';
DisplayPartKind.parameterName = 'parameterName';
DisplayPartKind.propertyName = 'propertyName';
DisplayPartKind.punctuation = 'punctuation';
DisplayPartKind.text = 'text';
export var SemicolonPreference;
(function (SemicolonPreference) {
    SemicolonPreference["Ignore"] = "ignore";
    SemicolonPreference["Insert"] = "insert";
    SemicolonPreference["Remove"] = "remove";
})(SemicolonPreference = SemicolonPreference || (SemicolonPreference = {}));
